colossalai/nn/layer/moe/experts.py

```diff
@@ -29,53 +35,48 @@ class Experts(MoeExperts):
     """
 
     def __init__(self, expert, num_experts, **expert_args):
-        super().__init__("all_to_all")
-
-        assert num_experts % moe_env.model_parallel_size == 0, \
-            "The number of experts should be divied by moe model size"
-
-        num_local_experts = num_experts // moe_env.model_parallel_size
+        super().__init__("all_to_all", num_experts)
 
-        with seed(ParallelMode.MOE_MODEL):
-            self.experts = nn.ModuleList([expert(**expert_args) for _ in range(num_local_experts)])
+        # Use seed to make every expert different from others
+        with seed(ParallelMode.TENSOR):
+            self.experts = nn.ModuleList([expert(**expert_args) for _ in range(self.num_local_experts)])
 
+        # Attach parallel information for all parameters in Experts
         for exp in self.experts:
             for param in exp.parameters():
-                param.__setattr__('moe_param', True)
-
-        self.num_local_experts = num_local_experts
+                param.__setattr__('moe_info', self.dist_info)

```

Comment:
Eevery param stores the duplicated dist_info?

Reply:
Python object only stores a reference. The dist_info is returned from moe_context information dictionary, there is no duplicated dist_info.

Reply:
I see